packages/process/src/node/terminal-process.ts

```diff
@@ -8,60 +8,75 @@
 import { injectable, inject } from 'inversify';
 import * as stream from 'stream';
 import { ILogger } from '@theia/core/lib/common';
-import { Process } from './process';
+import { Process, ProcessType } from './process';
 import { ProcessManager } from './process-manager';
-
-const pty = require("node-pty");
+import * as pty from 'node-pty';
+import { ITerminal } from 'node-pty/lib/interfaces';
 
 export const TerminalProcessOptions = Symbol("TerminalProcessOptions");
 export interface TerminalProcessOptions {
-    command: string,
-    args?: string[],
-    options?: object
+    readonly command: string,
+    readonly args?: string[],
+    readonly options?: object
 }
 
 export const TerminalProcessFactory = Symbol("TerminalProcessFactory");
-export type TerminalProcessFactory = (options: TerminalProcessOptions) => TerminalProcess;
+export interface TerminalProcessFactory {
+    (options: TerminalProcessOptions): TerminalProcess;
+}
 
 /* Use this instead of the node-pty stream, since the node-pty stream is already resumed.  */
-export class TerminalReadableStream extends stream.Readable {
-    constructor(protected readonly terminal: any, opts?: any) {
+class ReadableTerminalStream extends stream.Readable {
+
+    constructor(protected readonly terminal: ITerminal, opts?: stream.ReadableOptions) {
         super(opts);
-        this.terminal.on('data', (data: any) => {
-            this.push(data);
-        });
+        this.terminal.on('data', data => this.push(data));
     }
+
     /* This needs to be implemented as per node's API doc, even if it's empty.  */
     _read(size: number) {
     }
+
+}
+
+class WritableTerminalStream extends stream.Writable {
+
+    constructor(protected readonly terminal: ITerminal) {
+        super({
+            write: (chunk, encoding, next) => {
+                this.terminal.write(chunk.toString());
+                next();
+            }
+        });
+    }
+
```

Comment:
This should be removed as we discussed here: (https://github.com/theia-ide/theia/pull/841#issuecomment-344913607)  stdin is already in the pty we don't have to write it ourselves.